thirdparty/patchmatchnet/models/patchmatch.py

```diff
@@ -0,0 +1,951 @@
+                [self.dilation, -self.dilation],
+                [self.dilation, 0],
+                [self.dilation, self.dilation],
+            ]
+            for i in range(len(original_offset)):
+                offset_x, offset_y = original_offset[i]
+                original_offset.append([2 * offset_x, 2 * offset_y])
+        else:
+            raise NotImplementedError
+
+        with torch.no_grad():
+            y_grid, x_grid = torch.meshgrid(
+                [
+                    torch.arange(0, height, dtype=torch.float32, device=device),
+                    torch.arange(0, width, dtype=torch.float32, device=device),
+                ]
+            )
+            y_grid, x_grid = y_grid.contiguous(), x_grid.contiguous()
+            y_grid, x_grid = y_grid.view(height * width), x_grid.view(height * width)
+            xy = torch.stack((x_grid, y_grid))  # [2, H*W]
+            xy = torch.unsqueeze(xy, 0).repeat(batch, 1, 1)  # [B, 2, H*W]
+
+        xy_list = []
+        for i in range(len(original_offset)):
+            original_offset_y, original_offset_x = original_offset[i]
+
+            offset_x_tensor = original_offset_x + offset[:, 2 * i, :].unsqueeze(1)
+            offset_y_tensor = original_offset_y + offset[:, 2 * i + 1, :].unsqueeze(1)
+
+            xy_list.append((xy + torch.cat((offset_x_tensor, offset_y_tensor), dim=1)).unsqueeze(2))
+
+        xy = torch.cat(xy_list, dim=2)  # [B, 2, 9, H*W]
+
+        del xy_list, x_grid, y_grid
+
+        x_normalized = xy[:, 0, :, :] / ((width - 1) / 2) - 1
+        y_normalized = xy[:, 1, :, :] / ((height - 1) / 2) - 1
+        del xy
+        grid = torch.stack((x_normalized, y_normalized), dim=3)  # [B, 9, H*W, 2]
+        del x_normalized, y_normalized
+        grid = grid.view(batch, self.propagate_neighbors * height, width, 2)
+        return grid
+
+    def get_evaluation_grid(
+        self, batch: int, height: int, width: int, offset: torch.Tensor, device: torch.device, img: torch.Tensor = None
+    ) -> torch.Tensor:
+        """Compute the offests for adaptive spatial cost aggregation in adaptive evaluation
+
+        Args:
+            batch: batch size
+            height: grid height
+            width: grid width
+            offset: grid offset
+            device: decide the tensor on which device
+
+        Returns:
+            generated grid: in the shape of [batch, len(original_offset)*H, W, 2]
+        """
+        if self.evaluate_neighbors == 9:
+            dilation = self.dilation - 1  # dilation of evaluation is a little smaller than propagation
+            original_offset = [
+                [-dilation, -dilation],
+                [-dilation, 0],
+                [-dilation, dilation],
+                [0, -dilation],
+                [0, 0],
+                [0, dilation],
+                [dilation, -dilation],
+                [dilation, 0],
+                [dilation, dilation],
+            ]
+        elif self.evaluate_neighbors == 17:

```

Comment:
what does 17 or 9 represent here? can you explain that in the docstring or inline comment?